tools/clang/include/clang/Basic/Attr.td

```diff
@@ -984,6 +984,40 @@ def HLSLNodeTrackRWInputSharing : InheritableAttr {
   let Documentation = [Undocumented];
 }
 
+def HLSLNodeRecord : InheritableAttr {
+  let Spellings = []; // No spellings!
+  let Subjects = SubjectList<[CXXRecord]>;
+  let Documentation = [Undocumented];
+
+  let Args = [EnumArgument<"Type", "RecordType",
+                           ["DispatchNodeInputRecord",
+                           "RWDispatchNodeInputRecord",
+                           "GroupNodeInputRecords",
+                           "RWGroupNodeInputRecords",
+                           "ThreadNodeInputRecord",
+                           "RWThreadNodeInputRecord",
+                           "EmptyNodeInput",
+                           "NodeOutput",
+                           "EmptyNodeOutput",
+                           "NodeOutputArray",
+                           "EmptyNodeOutputArray",
+                           "GroupNodeOutputRecords",
+                           "ThreadNodeOutputRecords"],
+                           ["DispatchNodeInputRecord",
```

Comment:
This list and its formatting is ... unfortunate.  And it's ugly to have to define an enum locally for these, then translate between the local attribute's enum type and the `NodeIOKind` enum.  Not that I know of a better solution that works here, mainly because I am not familiar enough with what's available in this TableGen context.

Reply:
In upstream you can embed custom code into attribute definitions in tablegen, so the internal enum can be hidden entirely inside the attribute. I didn't want to make that large of a change here.

Reply:
Oh! I was actually wrong, it was just not hooked up where I expected it to be. Let me do a quick bit of cleanup...

Reply:
Ok, at least things that access the local enum are localized.  Still would be easy to make a mistake in lining up the enum strings/names with this formatting without easily seeing it in code review.